pkg/nextgen/distdoc/graph.go

```diff
@@ -0,0 +1,150 @@
+package distdoc
+
+import (
+	"container/list"
+	"errors"
+	"github.com/nuts-foundation/nuts-network/pkg/model"
+	"sort"
+)
+
+var errRootAlreadyExists = errors.New("root document already exists")
+
+// DAG is a directed acyclic graph consisting of nodes (documents) referring to preceding nodes.
+type DAG struct {
+	// root contains the root node of the DAG
+	root  *node
+	nodes map[string]*node
+	// missingEdges contains the forward references (A -> [B, C]) which couldn't be updated since documents were added out-of-order (B and C were added before A).
+	missingEdges map[string][]*node
+}
+
+type node struct {
+	document Document
+	// Documents 'prevs' field is a backreference (B -> A) so we store the forward reference (A -> B) so we can walk
+	// the DAG top to bottom.
+	edges map[string]*node
+}
+
+type Visitor func(document Document, next []Document)
```

Comment:
docs
why would a visitor need the next documents?

Reply:
Was actually there for the dot-visitor.  I changed its internals so it's not needed anymore.